### ISSUE TITLE
Clamp or repeat animation channels?

### ISSUE THREAD

**Issue by pjcozzi** [KhronosGroup/glTF#1179]
Moving spec discussion here from https://github.com/KhronosGroup/glTF-Sample-Models/issues/126.

If an animation has two channels of different length, what should the shorter channel once it finishes?
* Loop
* Clamp to the last keyframe
* Something else?

For example, if channel **a** is 2 seconds and channel **b** is 1 second, does **b** play twice or does it clamp to its last keyframe for the last second?

_Related_: do the channel lengths need to be evenly divisible, e.g., above, **b** could not be 1.5 seconds because that doesn't divide evenly into the longest channel time of 2?

Some use cases (like the Santa model) seem like looping is the right approach; other sample models (like the two boxes in https://github.com/KhronosGroup/glTF-Sample-Models/issues/126) seem like clamping is the right approach.

Does this need an extension to make this explicit?  Or limitations added to the spec, e.g., all channels must be the same length?  Or exporter/converter best practices?  Or something else?

Three.js, BabylonJS, and Cesium were all doing something different.

CC @bghgary @likangning93

**Comment by donmccurdy** [comment]
> Three.js, BabylonJS, and Cesium were all doing something different.

Three.js is looping each channel independently in the linked issue only because the channels are part of distinct animations, and each animation is treated as a "clip" or "action" per previous discussions. I think that we should continue to consider cross-animation timing out of scope for now.

If the two channels are part of the same clip, three.js will clamp the shorter animation. What were BabylonJS and Cesium doing? ([having trouble](https://github.com/BabylonJS/Babylon.js/issues/3386) with the babylon sandbox at the moment)

I would be inclined to vote:

1. No comment on multi-animation timing.
2. Channels within a single animation do not need to be the same length.
3. If channels within a single animation are not the same length, engine should clamp shorter animations to last frame.
4. Let's add exporter/converter best practices IFF engines will be inconsistent about this.

**Comment by donmccurdy** [comment]
Copying over @likangning93's comment from previous thread:

> Babylon "wraps" when time is outside a channel's given keyframes and Three.js holds the nearest keyframe, while Cesium can do either but defaults to holding.

When wrapping in Babylon, what happens if the entire clip is looped? Suppose an animation has two channels, 2s and 1.5s long respectively. If the channel is playing on a loop, does the 1.5s channel restart at the 2s mark, or only in 1.5s increments?

**Comment by bghgary** [comment]
> When wrapping in Babylon, what happens if the entire clip is looped?

BabylonJS loader treats each channel independently, so each channel will loop back to the beginning individually.

> does the 1.5s channel restart at the 1.5s and 2s marks, or only in 1.5s increments?

Only in 1.5s increments.

**Comment by najadojo** [comment]
Windows Mixed Reality implements as @donmccurdy describes:
- animations run independently 
- short channels within an animation clamp to the last frame

Looping isn't described the spec, we'd need to develop a description of what that means and how authors could control that behavior.

**Comment by javagl** [comment]
Referring to https://github.com/KhronosGroup/glTF/issues/573 , one might consider leaving this to the content creator, who can add a behavior enum: `ONCE`, `LOOP`, `PING_PONG` - but the linked issue referred to *animations* - I still have to wrap my head around whether something like this could make sense for the individual *channels* ...

**Comment by donmccurdy** [comment]
The three.js implementation is inspired quite heavily by Unity's animation system, in which an `animation` is analogous to an [AnimationClip](https://docs.unity3d.com/560/Documentation/Manual/class-AnimationClip.html). Loop behavior and clip duration (if different from span of the keyframes) can be customized at the AnimationClip level. I'm not familiar with Unreal but this looks comparable to its [Animation Sequences](https://docs.unrealengine.com/latest/INT/Engine/Animation/Sequences/).

> I still have to wrap my head around whether something like [ ONCE, LOOP, PING_PONG ] could make sense for the individual channels

These behaviors are more commonly used at the per-animation level, and having them on _both_ animations and channels sounds very complicated, so IMO we should avoid offering per-channel looping behavior. I'm also not sure how an artist would author animation intended to have channels loop independently... Typically in Blender or Maya you would set keyframes for each object or property, loop everything on a global timeline, and properties clamp before/after their first/last keyframes.

The alternative is to require that exporters/converters write duplicate keyframes to prevent this behavior, which is less efficient.

If [BabylonJS does not make a distinction between animations and channels](http://doc.babylonjs.com/babylon101/animations), then this may be a leaky comparison. Could the Babylon loader set the duration for each BABYLON.Animation to the duration of the longest channel in that animation, so that playback will align consistently? Or to copy the last keyframe while loading to the end of the clip?

**Comment by donmccurdy** [comment]
**tl;dr** I would still vote for the four [suggestions above](https://github.com/KhronosGroup/glTF/issues/1179#issuecomment-351436939). Whether we want an implementation note to converters/exporters may depend on whether we can change this behavior in Babylon... would be nice to not recommend sending duplicate animation data as best practice if there's any way we can make the implementations consistent.

**Comment by bghgary** [comment]
I talked to tech art and here is some info:
- Artists tools (like Maya) deal with a single timeline for animations. The timeline can have multiple animation curves targeting different properties that execute in parallel. Artists typically author multiple animation clips within one timeline and export each animation clip separately (say to an FBX). There is no concept of a clip or a subsection of the timeline. The ranges of clips are typically handled by the consuming engine. For example, in Unity, the clip is defined on import.
- When looping, the clamping behavior for short curves that don't span the entire timeline is expected in all the artist tools that we are aware of.

Given this, I believe BabylonJS implementation is doing this wrong and needs to be fixed. Implementation note in the spec might be helpful, but I don't think it's required.

> Could the Babylon loader set the duration for each BABYLON.Animation to the duration of the longest channel in that animation, so that playback will align consistently? Or to copy the last keyframe while loading to the end of the clip?

I have been looking into how to fix this. I think what I want to do is implement something for BabylonJS runtime that groups BABYLON.Animation objects (which are single curves) together into one animation and then handle the union of the time range and clamping behavior, but I'll need to discuss this with the team.

Related question: The end times don't have to match for channels of a single animation, but do the start times have to match? Doesn't seem like it from the spec language. If so, then the clamping behavior should apply to both ends of the curve.

**Comment by bghgary** [comment]
> Some use cases (like the Santa model) seem like looping is the right approach; other sample models (like the two boxes in KhronosGroup/glTF-Sample-Models#126) seem like clamping is the right approach.

From an authoring perspective, the typical approach for doing something like the Santa model is to export two separate animation clips and play them together. That isn't possible yet with glTF because the spec doesn't describe how glTF animations (i.e. animation clips) should be played. I suspect implementations will default to playing just the first animation. Short-term, for the Santa model, the way to get it to work properly exported to glTF is to combine the curves into a single glTF animation and repeat the short curve manually in the timeline until it is as long as the second curve.

**Comment by donmccurdy** [comment]
> There is no concept of a clip or a subsection of the timeline. The ranges of clips are typically handled by the consuming engine. For example, in Unity, the clip is defined on import.

That sounds right. Importing multiple FBX files, each with a copy of the mesh and a different animation, is obviously not ideal for a runtime format... so looking closely at the end result in Unity seems like a better option. 

> Related question: The end times don't have to match for channels of a single animation, but do the start times have to match? Doesn't seem like it from the spec language. If so, then the clamping behavior should apply to both ends of the curve.

Agreed. I do not think the start times should need to match either, so clamping should be applied in both cases.

**Comment by pjcozzi** [comment]
>>Some use cases (like the Santa model) seem like looping is the right approach; other sample models (like the two boxes in KhronosGroup/glTF-Sample-Models#126) seem like clamping is the right approach.

>From an authoring perspective, the typical approach for doing something like the Santa model is to export two separate animation clips and play them together. That isn't possible yet with glTF because the spec doesn't describe how glTF animations (i.e. animation clips) should be played. I suspect implementations will default to playing just the first animation. Short-term, for the Santa model, the way to get it to work properly exported to glTF is to combine the curves into a single glTF animation and repeat the short curve manually in the timeline until it is as long as the second curve.

Perhaps this indicates
* Changes are needed to COLLADA2GLTF, which generated the Santa model
* Exporter best practices documentation is needed

**Comment by donmccurdy** [comment]
> * Changes are needed to COLLADA2GLTF, which generated the Santa model

^ That's partly handled by https://github.com/KhronosGroup/COLLADA2GLTF/pull/102. If the Santa COLLADA model contains any indication of looping or `<animation_clip/>` nodes, that would probably indicate the next step.

**Comment by bghgary** [comment]
> @bghgary were you able to update the BabylonJS implementation?

Yes, this should be fixed in https://github.com/BabylonJS/Babylon.js/pull/3554 as referenced.

**Comment by donmccurdy** [comment]
Great, thanks! Closing this out. I'm not sure that an implementation note is necessary to explain clamping channels outside their defined input ranges, but glad to consider it if someone feels strongly and can propose wording.
